babel.config.js

```diff
@@ -25,99 +25,394 @@
+  const depCheckPlugins = [
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-classes'), {loose: false}],
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
   ];
-  const transformClasses = [
-    '@babel/plugin-transform-classes',
-    {
-      loose: false,
-    },
+  const depCheckPresets = [defaultPresetEnvPlugin];
+  return {
+    compact: false,
+    ignore: ignoredGlobalModules,
+    plugins: depCheckPlugins,
+    presets: depCheckPresets,
+    sourceType: 'module',
+  };
+}
+
+/**
+ * Gets the config for babel transforms run during `gulp build`.
+ *
+ * @return {!Object}
+ */
+function getUnminifiedConfig() {
+  const unminifiedPlugins = [
+    replacePlugin,
+    localPlugin('transform-json-configuration'),
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-classes'), {loose: false}],
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
   ];
+  const unminifiedPresets = [defaultPresetEnvPlugin];
+  return {
+    compact: false,
+    ignore: ignoredGlobalModules,
+    plugins: unminifiedPlugins,
+    presets: unminifiedPresets,
+    sourceType: 'module',
+  };
+}
 
-  if (esm) {
-    return [leadingComments, reactConstantElements, transformJSX];
+/**
+ * Gets the config for pre-closure babel transforms run during `gulp dist`.
+ *
+ * @return {!Object}
+ */
+function getPreClosureConfig() {
+  const preClosurePlugins = [
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
+    localPlugin('transform-inline-configure-component'),
+    // TODO(alanorozco): Remove `replaceCallArguments` once serving infra is up.
+    [localPlugin('transform-log-methods'), {replaceCallArguments: false}],
+    localPlugin('transform-parenthesize-expression'),
+    localPlugin('is_minified-constant-transformer'),
+    localPlugin('transform-amp-extension-call'),
+    localPlugin('transform-html-template'),
+    localPlugin('transform-version-call'),
+    localPlugin('transform-simple-array-destructure'),
+    replacePlugin,
+  ];
+  if (argv.single_pass) {
+    preClosurePlugins.push(localPlugin('transform-amp-asserts'));
+  }
+  if (argv.esm) {
+    preClosurePlugins.push(
+      defaultFilterImportsPlugin,
+      localPlugin('transform-function-declarations')
+    );
+  }
+  const isCheckTypes = argv._.includes('check-types');
+  if (isCheckTypes) {
+    preClosurePlugins.push(localPlugin('transform-simple-object-destructure'));
+  } else {
+    preClosurePlugins.push(localPlugin('transform-json-configuration'));
+  }
+  if (!argv.fortesting && !isCheckTypes) {
+    preClosurePlugins.push(
+      [localPlugin('amp-mode-transformer'), {isEsmBuild: argv.esm}],
+      localPlugin('is_dev-constant-transformer')
+    );
   }
+  const babelPresetEnvOptions = argv.esm
```

Comment:
It's a bit confusing to orchestrate the preset-env options here and also in a dedicated helper method at the top of this module.

If we add a new flag to babel preset-env, it seems likely we'll miss it in one of these three places. (Note, not sure of a better solution, just wanted to point this out for you to consider).

Reply:
Resolved this by moving configs to different files and getting rid of the `const`s at the top of the file.